modules/core/src/test/java/org/apache/ignite/internal/processors/cache/persistence/freelist/FreeListTest.java

```diff
@@ -0,0 +1,273 @@
+     * With various garbage collectors and with assertions both turned on and off.
+     */
+    @Test
+    public void testCutTail() throws Exception {
+        try (Ignite ignite = prepareCluster(jvmOpts,
+            "org.apache.ignite.internal.processors.cache.persistence.freelist.PagesList$CutTail::run")) {
+            IgniteFuture<Void> jobFut = ignite.compute().runAsync(new TestCutTailJob());
+
+            assertTrue(c2JitCompiled.await(getTestTimeout() / 2, TimeUnit.MILLISECONDS));
+
+            try {
+                // Let it work for 2 seconds with the optimized C2 Jit-compiled variant of the
+                // PagesList$CutTail::run method. It's enough to reproduce problem if it is broken.
+                jobFut.get(2000, TimeUnit.MILLISECONDS);
+            }
+            catch (IgniteFutureTimeoutException e) {
+                jobFut.cancel();
+            }
+        }
+    }
+
+    /**
+     * Starts server node in remote JVM with the JVM options passed.
+     * Registers listener to signal once the method passed is C2 Jit-compiled in server node.
+     * Starts client node in local JVM.
+     *
+     * @param jvmOpts JVM options for the server node.
+     * @param method Fully qualified name of method.
+     * @return Client node.
+     */
+    private Ignite prepareCluster(List<String> jvmOpts, String method) throws Exception {
+        CountDownLatch remoteJvmServerStarted = new CountDownLatch(1);
+
+        ListeningTestLogger lsnrLog = new ListeningTestLogger(log);
+
+        lsnrLog.registerListener(new CallbackExecutorLogListener(".*Topology snapshot \\[ver=1,.*",
+            remoteJvmServerStarted::countDown));
+
+        lsnrLog.registerListener(new CallbackExecutorLogListener(
+            ".*Compiled method \\(c2\\).*" + method.replace("$", "\\$") + ".*",
+            c2JitCompiled::countDown));
+
+        IgniteConfiguration cfg = optimize(getConfiguration("remote-jvm-server"));
+
+        new IgniteProcessProxy(cfg, lsnrLog, null, false) {
+            @Override protected Collection<String> filteredJvmArgs() throws Exception {
+                Collection<String> args = super.filteredJvmArgs();
+
+                args.remove("-ea");
+
+                args.add("-XX:+UnlockDiagnosticVMOptions");
+                args.add("-XX:PrintAssemblyOptions=intel");
+                args.add("-XX:CompileCommand=print," + method);
+
+                args.addAll(jvmOpts);
+
+                return args;
+            }
+        };
+
+        remoteJvmServerStarted.await(getTestTimeout(), TimeUnit.MILLISECONDS);
+
+        return startClientGrid("local-jvm-client");
+    }
+
+    /**
+     * Perfrom concurrent updates and deletes ensuring the CutTail called enough
+     * times to invoke the C2 optimizing Jit compiler.
+     */
+    private static class TestCutTailJob implements IgniteRunnable {
+        /** */
+        @IgniteInstanceResource
+        Ignite ignite;
+
+        /** */
+        Random random = new Random(3793);
```

Comment:
Magic number. 
Why not use Random() without seed?
Perhaps ThreadLocalRandom?

Reply:
Just to reduce randomness. Besides looks like contention on global Random increased chances to reproduce the problem.

Now with the introduction of explicit barrier it's not needed any more.

Will replace with ThreadLocalRandom.